dali/kernels/common/transpose.h

```diff
@@ -0,0 +1,214 @@
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+
+namespace dali {
+namespace kernels {
+
+namespace detail {
+
+template <typename T, int LevelsLeft, int MaxLevels = -1>
+std::enable_if_t<LevelsLeft == 1> transpose_impl_static(T *dst, const T *src, int max_levels,
+                                                        span<const int64_t> dst_stride,
+                                                        span<const int64_t> src_stride,
+                                                        TensorShape<> size, span<const int> perm) {
+  const auto level = MaxLevels >= 0 ? (MaxLevels - LevelsLeft) : (max_levels - LevelsLeft);
+  auto dst_level_stride = dst_stride[level];
+  auto src_level_stride = src_stride[perm[level]];
+  for (int64_t i = 0; i < size[level]; i++) {
+    *dst = *src;
+    dst += dst_level_stride;
+    src += src_level_stride;
+  }
+}
+
+template <typename T, int LevelsLeft, int MaxLevels = -1>
+std::enable_if_t<(LevelsLeft > 1)> transpose_impl_static(T *dst, const T *src, int max_levels,
+                                                         span<const int64_t> dst_stride,
+                                                         span<const int64_t> src_stride,
+                                                         TensorShape<> size, span<const int> perm) {
+  const auto level = MaxLevels >= 0 ? (MaxLevels - LevelsLeft) : max_levels - LevelsLeft;
+  auto dst_level_stride = dst_stride[level];
+  auto src_level_stride = src_stride[perm[level]];
+  for (int64_t i = 0; i < size[level]; i++) {
+    transpose_impl_static<T, LevelsLeft - 1, MaxLevels>(dst, src, max_levels, dst_stride,
+                                                        src_stride, size, perm);
+    dst += dst_level_stride;
+    src += src_level_stride;
+  }
+}
+
+template <typename T, int Static = 3>
+void transpose_impl(T *dst, const T *src, int level, int max_levels, span<const int64_t> dst_stride,
+                    span<const int64_t> src_stride, TensorShape<> size, span<const int> perm) {
+  auto dst_level_stride = dst_stride[level];
+  auto src_level_stride = src_stride[perm[level]];
+
+  if (max_levels - level == Static) {
+    transpose_impl_static<T, Static>(dst, src, max_levels, dst_stride, src_stride, size, perm);
+  } else {
+    for (int64_t i = 0; i < size[level]; i++) {
+      transpose_impl<T, Static>(dst, src, level + 1, max_levels, dst_stride, src_stride, size,
+                                perm);
+      dst += dst_level_stride;
+      src += src_level_stride;
+    }
+  }
+}
+
+/**
+ * @brief Find blocks of consecutive numbers in the permutation that would be transposed together
+ *
+ * For example [0, 1, 5, 2, 3, 4] has following blocks: [[0, 1], [5], [2, 3, 4]]
+ * and the result would be: {{0, 1}, {2, 2}, {3, 5}}
+ *
+ * @return List of starting and ending position of blocks(both ends inclusive)
+ */
+std::vector<std::pair<int, int>> PermutationBlocks(span<const int> perm) {
```

Comment:
Returning std::vector looks appealing, but it's rather costly.
I'd recommend returning a SmallVector of reasonable static size (6?).

Reply:
Done